filters/auth/oauth.go

```diff
@@ -432,4 +585,238 @@ func (f *filter) Request(ctx filters.FilterContext) {
 	ctx.StateBag()[tokeninfoCacheKey] = authMap
 }
 
-func (f *filter) Response(filters.FilterContext) {}
+func (f *tokeninfoFilter) Response(filters.FilterContext) {}
+
+// NewOAuthTokenintrospectionAnyKV creates a new auth filter specification
+// to validate authorization for requests. Current implementation uses
+// Bearer tokens to authorize requests and checks that the token
+// contains at least one key value pair provided.
+//
+// This is implementing RFC 7662 compliant implementation. It uses
+// POST requests to call introspection_endpoint to get the information
+// of the token validity.
+//
+// It uses /.well-known/openid-configuration path to the passed
+// oauthIssuerURL to find introspection_endpoint as defined in draft
+// https://tools.ietf.org/html/draft-ietf-oauth-discovery-06, if
+// oauthIntrospectionURL is a non empty string, it will set
+// IntrospectionEndpoint to the given oauthIntrospectionURL.
+func NewOAuthTokenintrospectionAnyKV(oauthIssuerURL, oauthIntrospectionURL string) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAnyKV, oauthIssuerURL, oauthIntrospectionURL)
+}
+
+// NewOAuthTokenintrospectionAllKV creates a new auth filter specification
+// to validate authorization for requests. Current implementation uses
+// Bearer tokens to authorize requests and checks that the token
+// contains at least one key value pair provided.
+//
+// This is implementing RFC 7662 compliant implementation. It uses
+// POST requests to call introspection_endpoint to get the information
+// of the token validity.
+//
+// It uses /.well-known/openid-configuration path to the passed
+// oauthIssuerURL to find introspection_endpoint as defined in draft
+// https://tools.ietf.org/html/draft-ietf-oauth-discovery-06, if
+// oauthIntrospectionURL is a non empty string, it will set
+// IntrospectionEndpoint to the given oauthIntrospectionURL.
+func NewOAuthTokenintrospectionAllKV(oauthIssuerURL, oauthIntrospectionURL string) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAllKV, oauthIssuerURL, oauthIntrospectionURL)
+}
+
+func NewOAuthTokenintrospectionAnyClaims(oauthIssuerURL, oauthIntrospectionURL string) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAnyClaims, oauthIssuerURL, oauthIntrospectionURL)
+}
+func NewOAuthTokenintrospectionAllClaims(oauthIssuerURL, oauthIntrospectionURL string) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAllClaims, oauthIssuerURL, oauthIntrospectionURL)
+}
+
+func newOAuthTokenintrospectionFilter(typ roleCheckType, oauthIssuerURL, oauthIntrospectionURL string) filters.Spec {
+	cfg, err := getOpenIDConfig(oauthIssuerURL)
+	if err != nil {
+		return &tokenIntrospectionSpec{
```

Comment:
We should move the open-id-config file checking/reading to the calling code (the root package), and here only use the instrospection url from it. There, only the case when the file doesn't exist should be ignored, otherwise the startup should fail.

Reply:
update: actually, if the calling code specifies the `oauthIntrospectionURL` explicitly, then we probably should call the getOpenIDConfig at all. What do you think?

Reply:
see below at the dispatch of f.typ to checkOAuthTokenintrospectionAnyClaims, it's checking s.config.ClaimsSupported, which is only got from the open-id-config.

On the other hand @vetinari asked me why not to use https://github.com/coreos/go-oidc ?
I see some problems with that library, but I want to try it before I do not consider it. It can more easily provide some nice flow integrations, not sure if this is easily possible in our case, but let's see.

Reply:
i see. let's try it.

i would still try to find a way to return the error during the startup. I see that the filter spec creation in the builtin.MakeRegistry() doesn't allow returning errors, which I guess was the main problem. Maybe, if getOpenIDConfig or the go-oidc sutff is called from outside the NewOAuth... functions, in the root package, and the available options are passed to the NewOAuth... functions with the available info, including the open id config, it is possible ensure failing early.

Reply:
Yes exactly, maybe this makes sense. I am also not sure if we want to poll for updates of this configuration from time to time. Right now I don't know and maybe this means pulling it into the skipper bootstrapping process makes most sense than overcomplicate this.